dvc/tree/dropbox.py

```diff
@@ -0,0 +1,466 @@
+
+    @abstractmethod
+    def save(self, creds):
+        raise NotImplementedError
+
+
+class CliCredProvider(DropboxCredProvider):
+    def can_auth(self):
+        return True
+
+    def _gather_auth(self):
+        import dropbox
+
+        flow = dropbox.DropboxOAuth2FlowNoRedirect(
+            consumer_key=self.DROPBOX_APP_KEY,
+            consumer_secret=self.DROPBOX_APP_SECRET,
+            token_access_type="offline",
+        )
+        authorize_url = flow.start()
+        print("Go to the following link in your browser:")
+        print()
+        print("    " + authorize_url)
+        print()
+        code = input("Enter verification code: ").strip()
+        try:
+            auth = flow.finish(code)
+        except dropbox.oauth.NotApprovedException as ex:
+            raise DvcException(
+                "To use Dropbox remote, you need to approve the DVC "
+                "application to access your Dropbox account.\n"
+                "Learn more at {}".format(
+                    format_link("https://man.dvc.org/remote/modify")
+                )
+            ) from ex
+        return {
+            ACCESS_TOKEN: auth.access_token,
+            EXPIRATION: auth.expires_at.isoformat(),
+            REFRESH_TOKEN: auth.refresh_token,
+        }
+
+    def save(self, creds):
+        pass
+
+
+class EnvCredProvider(DropboxCredProvider):
+    def can_auth(self):
+        return bool(os.environ.get(REFRESH_TOKEN, False))
+
+    def _gather_auth(self):
+        return {
+            ACCESS_TOKEN: os.environ.get(ACCESS_TOKEN, ""),
+            EXPIRATION: os.environ.get(EXPIRATION, ""),
+            REFRESH_TOKEN: os.environ[REFRESH_TOKEN],
+        }
+
+    def save(self, creds):
+        _creds = creds.copy()
+        _creds[EXPIRATION] = _creds[EXPIRATION].isoformat()
+        os.environ.update(_creds)
+
+
+class FileCredProvider(DropboxCredProvider):
+    DEFAULT_FILE = "dropbox-user-credentials.json"
+    DROPBOX_CREDENTIALS_FILE = "DROPBOX_CREDENTIALS_FILE"
+
+    def __init__(self, repo):
+        self._cred_location = os.environ.get(
+            self.DROPBOX_CREDENTIALS_FILE,
+            os.path.join(repo.tmp_dir, self.DEFAULT_FILE),
+        )
+
+    def can_auth(self):
+        if not os.path.exists(self._cred_location):
+            return False
+        with open(self._cred_location) as infile:
+            return REFRESH_TOKEN in json.load(infile)
+
+    def _gather_auth(self):
+        with open(self._cred_location) as infile:
+            config = json.load(infile)
+        return {
+            ACCESS_TOKEN: config.get(ACCESS_TOKEN, ""),
+            EXPIRATION: config.get(EXPIRATION, ""),
+            REFRESH_TOKEN: config[REFRESH_TOKEN],
+        }
+
+    def save(self, creds):
+        _creds = creds.copy()
+        _creds[EXPIRATION] = _creds[EXPIRATION].isoformat()
+        with open(self._cred_location, "w") as outfile:
+            json.dump(_creds, outfile)
+
+
+def path_info_to_dropbox_path(path_info):
```

Comment:
I've updated the code in https://github.com/gmrukwa/dvc/pull/1 to allow empty hostname so that we don't have to deal with `bucket` that is not bucket at all (and to deal with cases like `import-url dropbox:///file.pdf`

still, I wonder what is the best way to handle this leading `/` here, @efiop ? Dropbox expects them this way? Is there a convenience method in PathInfo to return path with `/`?

Reply:
Dropbox generally expects a path that is like `/a/path/to/file`, which gets rooted at the root of your Dropbox home. I couldn't get it to work simply without such a construct unfortunately. You will rather know some better way.

Reply:
yep, I did some changes to support URL without hostname (which is not present here). Usually the look like `file:///path` (mind triple `///`).